05-java-ios/COURSE_MATERIAL.md

```diff
@@ -64,26 +73,321 @@ without the hassle of network programming.
+`BinaryBufferReadFileExample.java` file:
+
+```diff
+   public static void main(String[] args) throws IOException {
+     InputStream fis = new FileInputStream("binary-file.bin");
++    InputStream bis = new BufferedInputStream(fis);
+
+     // -1 indicates the end of the file
+     int b;
+-    while ((b = fis.read()) != -1) {
++    while ((b = bis.read()) != -1) {
+       System.out.print(b);
+     }
+
++    bis.close();
+     fis.close();
+   }
+```
+
+The `BufferedInputStream` class is created with the `FileInputStream` class as
+an argument. The `BufferedInputStream` class will read data from the
+`FileInputStream` class and store it in a buffer.
+
+The `BufferedInputStream` class will read data from the buffer instead of
+reading data from the file directly. This is more efficient as the buffer can
+store more data than the `FileInputStream` class can read in one system call.
 
-Computers only understand binary data. Binary data is a sequence of bits (`0`s
-and `1`s). Binary data can represent anything: text, images, videos, etc. On the
-file system, everything is binary data.
+Compile and execute the `BinaryBufferReadFileExample.java` file. This will read
+the content of the file `binary-file.bin` and print the file content to the
+console.
 
-The real question is: how do we interpret these bits?
+Open the `BinaryBufferWriteFileExample.java` file in the `05-java-ios` directory
+to see how to write binary data with a buffer.
+
+Here are the changes between the `BinaryWriteFileExample.java` file and the
+`BinaryBufferWriteFileExample.java` file:
+
+```diff
+   public static void main(String[] args) throws IOException {
+     OutputStream fos = new FileOutputStream("binary-file.bin");
++    BufferedOutputStream bos = new BufferedOutputStream(fos);
+
+     for (int i = 0; i < 256; i++) {
+-      fos.write(i);
++      bos.write(i);
+     }
+
++    // Flush the buffer to write the remaining bytes as the buffer may not be full
++    bos.flush();
++    bos.close();
+     fos.close();
+   }
+```
+
+The `BufferedOutputStream` class is created with the `FileOutputStream` class as
+an argument. The `BufferedOutputStream` class will write data to the
+`FileOutputStream` class and store it in a buffer.
+
+The `BufferedOutputStream` class will write data to the buffer instead of
+writing data to the file directly. This is more efficient as the buffer can
+store more data than the `FileOutputStream` class can write in one system call.
+
+The `flush()` method is called to write the remaining bytes in the buffer to the
+file. This is important to call the `flush()` method to make sure that all data
+is written to the file.
+
+Compile and execute the `BinaryBufferWriteFileExample.java` file. This will
+create the file `binary-file.bin` in the current directory.
+
+Compile and execute the `BinaryBufferReadFileExample.java` file. This will read
+the content of the file `binary-file.bin` and print the file content to the
+console.
+
```

Comment:
After reading the whole chapter, I think it's very well rephrased. However, I would add a few details about buffers to make it clearer.

- When reading for the first time, a system call is made and the data read is stored in the buffer.
- Then, as long as the buffer is not empty, we read into it (no syscall).
- As soon as the buffer is empty, a new syscall is made and the process is repeated.

The same applies to writing.

This information has already been more or less said, but I think it would be good to be quite clear about how it works. 

We can discuss it if you like